src/proof/lrat/lrat_proof.h

```diff
@@ -0,0 +1,207 @@
+/*********************                                                        */
+/*! \file lrat_proof.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief LRAT Proof Format
+ **
+ ** Declares C++ types that represent a LRAT proof.
+ ** Defines serialization for these types.
+ **
+ ** Represents an **abstract** LRAT proof.
+ ** Does **not** represent an LFSC LRAT proof, or an LRAT proof being used to
+ ** prove things about bitvectors.
+ **
+ ** Paper on LRAT: https://www.cs.utexas.edu/~marijn/publications/lrat.pdf
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__PROOF__LRAT__LRAT_PROOF_H
+#define __CVC4__PROOF__LRAT__LRAT_PROOF_H 0
+
+#include <vector>
+#include <unordered_map>
+#include <string>
+
+#include "base/output.h"
+#include "proof/clause_id.h"
+// Included because we nee operator<< for the SAT types
+#include "prop/sat_solver.h"
+
+namespace CVC4 {
+namespace proof {
+namespace lrat {
+
+// Refers to clause position within an LRAT proof
+using ClauseIdx = size_t;
+
+enum LRATInstructionKind
+{
+  lratDeletion,
+  lratAddition,
+};
+
+struct LRATDeletionData
+{
+  // This idx doesn't really matter, but it's in the format anyway, so we parse
+  // it.
+  ClauseIdx idxOfClause;
+
+  // Clauses to delete
+  std::vector<ClauseIdx> clauses;
+};
+
+// A sequence of locations that will contain unit clauses during unit
+// propegation
+using LRATUPTrace = std::vector<ClauseIdx>;
+
+struct LRATAdditionData
+{
+  // The idx for the new clause
+  ClauseIdx idxOfClause;
+  // The new clause
+  prop::SatClause clause;
+  // UP trace based on the negation of that clause
+  LRATUPTrace atTrace;
+
+  // Clauses that can resolve with `clause` on its first variable,
+  // together with a UP trace after that resolution.
+  // Used for RAT checks.
+  std::vector<std::pair<ClauseIdx, LRATUPTrace>> resolvants;
+};
+
+// This is conceptually an Either<Addition,Deletion>
+struct LRATInstruction
```

Comment:
My reason for choosing a disjoint union over an inheritance hierarchy was that I thought it would produce more understandable code.  In particular, when implementing a particular **verb** for an inheritance hierarchy, the code get's split up, while the implementation  of a particular **verb** for a disjoint union is all in one places.

That being said, while writing this code I didn't (yet) know how to implement disjoin  unions correctly in C++. Now I do, so I'll clean it up. Take a look at the result, and if you'd like, I'm happy to continue the conversation.